src/bluesky/tests/test_flyer.py

```diff
@@ -4,17 +4,80 @@
 import pytest
 from ophyd import Component as Cpt
 from ophyd import Device
-from ophyd.sim import NullStatus, TrivialFlyer
+from ophyd.sim import NullStatus, StatusBase, TrivialFlyer
 
 from bluesky import Msg
-from bluesky.plan_stubs import close_run, complete, complete_all, kickoff, kickoff_all, open_run, wait
+from bluesky.plan_stubs import (
+    close_run,
+    collect_while_completing,
+    complete,
+    complete_all,
+    declare_stream,
+    kickoff,
+    kickoff_all,
+    open_run,
+    wait,
+)
 from bluesky.plans import count, fly
 from bluesky.protocols import Preparable
 from bluesky.run_engine import IllegalMessageSequence
 from bluesky.tests import requires_ophyd
 from bluesky.tests.utils import DocCollector
 
 
+def collect_while_completing_plan(flyers, dets, stream_name: str = "test_stream", pre_declare: bool = True):
+    yield from open_run()
+    if pre_declare:
+        yield from declare_stream(*dets, name=stream_name, collect=True)
+    yield from collect_while_completing(flyers, dets, flush_period=0.1, stream_name=stream_name)
+    yield from close_run()
+
+
+def test_collect_while_completing_plan_trivial_case(RE):
+    completed = []
+    collected = []
+
+    class StatusDoneAfterSecondCall(StatusBase):
+        times_called = 0
+
+        @property
+        def done(self):
+            self.times_called += 1
+            return self.times_called >= 10
+
+    class SlowFlyer:
+        name = "trivial-flyer"
+        custom_status = StatusDoneAfterSecondCall()
+
+        def kickoff(self):
+            return NullStatus()
+
+        def complete(self):
+            completed.append(self)
+            return self.custom_status
+
+    class TrivialDetector:
+        name = "trivial-detector"
+
```

Comment:
Please could you add a `collect` method to this that makes an incrementing number each time it is called, then check that you get the right number of `Event` documents out of the RE by adding a callback.

Something along the lines of this might help:
```python
from collections import defaultdict

def assert_emitted(docs: Dict[str, list], **numbers: int):
    assert list(docs) == list(numbers)
    assert {name: len(d) for name, d in docs.items()} == numbers

def test_plan():
    docs = defaultdict(list)
    RE(plan(), lambda name, doc: docs[name].append(doc))
    assert_emitted(docs, start=1, descriptor=1, resource=1, datum=1, event=1, stop=1)
```